### PR TITLE
C# tools: support generated filename corner cases

### LINKED ISSUE #17661: Grpc.Tools causes a build error when .proto file name contains dot
### What version of gRPC and what language are you using?
 
Grpc 1.17.1
Grpc.Core 1.17.1
Grpc.Tools 1.17.1

Language: C#
 
### What operating system (Linux, Windows, …) and version?
 
Windows 10 Pro 1803 64bit
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
.NET Core SDK 2.1.502
 
### What did you do?

I tried to build .proto  which file name contains dot(e.g. hello.world.proto) by Grpc.Tools.

source files: [Greeter.zip](https://github.com/grpc/grpc/files/2737695/Greeter.zip)
 
### What did you expect to see?
 
Build succeeds.
 
### What did you see instead?

Following build errors:
```
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(18,57,18,69): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(19,57,19,67): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(21,53,21,65): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(21,86,21,96): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(43,117,43,129): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(43,77,43,87): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(81,80,81,92): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(81,41,81,51): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(91,80,91,92): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(91,41,91,51): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(103,107,103,119): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(103,62,103,72): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(113,107,113,119): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
```

### PR DESCRIPTION
**Pull request by kkm000**
protoc and gRPC codegens differently treat non-ASCII letter characters and symbols other than underscores when constructing their respective output filenames for generated .cs files. This change reproduces their respective behaviors exactly.

I added test cases similar to my hello.world.proto flop, and verified the built dll by dropping it into the .nuget directory on a Linux machine. Build failed with a repro of #17661 before, succeeded after.

Closes #17661

### CODE REVIEW

File: src/csharp/Grpc.Tools/GeneratorServices.cs [hunk 1]
```diff
@@ -109,6 +108,26 @@ string LowerUnderscoreToUpperCamel(string str)
             }
             return result.ToString();
         }
+
+        // This is how the protoc codegen constructs its output filename.
+        // See protobuf/compiler/csharp/csharp_helpers.cc:356.
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/GeneratorServices.cs]: for comparison,  the logic is here:
https://github.com/protocolbuffers/protobuf/blob/3a3956e8a258784461270961c6577341356bce52/src/google/protobuf/compiler/csharp/csharp_helpers.cc#L137  with `cap_next_letter=true` and `preserve_period=false`

Comment (kkm000) [reply src/csharp/Grpc.Tools/GeneratorServices.cs]: Done.

File: src/csharp/Grpc.Tools/GeneratorServices.cs [hunk 2]
```diff
@@ -109,6 +108,26 @@ string LowerUnderscoreToUpperCamel(string str)
...
+            bool cap = true;
+            foreach (char c in str)
+            {
+                char upperC = char.ToUpperInvariant(c);
+                bool isAsciiLetter = 'A' <= upperC && upperC <= 'Z';
+                if (isAsciiLetter || ('0' <= c && c <= '9'))
+                {
+                    result.Append(cap ? upperC : c);
+                }
+                cap = !isAsciiLetter;
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/GeneratorServices.cs]: I believe there's still a subtle problem - after a digit, protoc will still capitalize the next letter:
I tested and `one123two.cs` generates `One123Two.cs` and `One123twoGrpc.cs`.

Comment (kkm000) [reply src/csharp/Grpc.Tools/GeneratorServices.cs]: Uhhhh, thanks! I'll take care of this.

Comment (kkm000) [reply src/csharp/Grpc.Tools/GeneratorServices.cs]: This passess, I added a test and checked in real build, too. `cap = !isAsciiLetter;` will be true after a digit, as well as after any (non-emitted) other character. I'll push the test case tho, so it exists.

Comment (kkm000) [reply src/csharp/Grpc.Tools/GeneratorServices.cs]: Done, aa40424.

Comment (jtattermusch) [reply src/csharp/Grpc.Tools/GeneratorServices.cs]: ah, thanks for clarification!

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/18477
https://github.com/grpc/grpc/issues/13379

**Pull request merged**